app/models/task/late_generator.rb

```diff
@@ -0,0 +1,63 @@
+require 'csv'
+
+class Task
+  # Used to generate a 'late' CSV for all users
+  # of suppliers with incomplete submissions in the
+  # given year/month period. Outputs via +puts+ objects
+  # that respond_to? it (+STDOUT+ or +File+ being usual)
+  class LateGenerator
+    HEADER = [
+      'User Name',
+      'Email Address',
+      'Framework Number'
+    ].freeze
+
+    attr_reader :logger, :output, :month, :year
+
+    def initialize(month:, year:, output: STDOUT, logger: Rails.logger)
+      @month = month
+      @year = year
+      @output = output
+      @logger = logger
+    end
+
+    delegate :info, :warn, to: :logger
+
+    ##
+    # NB should become its own class if this becomes an
+    # admin front end confection. This class should be concerned
+    # with generating CSV, not getting the records
+    def user_framework_lates
+      Task
+        .incomplete
+        .where(period_month: month, period_year: year)
+        .select('
+          tasks.id,
+          frameworks.short_name AS framework_short_name,
+          users.name AS user_name,
+          users.email AS user_email
+        ')
+        .joins(
+          :framework,
+          supplier: { memberships: :user }
+        )
+    end
+
+    def generate
+      logger.info "Generating late contacts for #{year}, #{month}"
+
+      output.puts(CSV.generate_line(HEADER))
+      user_framework_lates.each do |user_framework|
```

Comment:
Approving but one last suggestion: perhaps for_each here to avoid future memory bloat?